agenthub/codeact_agent/codeact_agent.py

```diff
@@ -194,27 +215,148 @@ def step(self, state: State) -> Action:
         if latest_user_message and latest_user_message.strip() == '/exit':
             return AgentFinishAction()
 
-        # prepare what we want to send to the LLM
-        messages: list[dict[str, str]] = self._get_messages(state)
-
-        response = self.llm.completion(
-            messages=messages,
-            stop=[
-                '</execute_ipython>',
-                '</execute_bash>',
-                '</execute_browse>',
-            ],
-            temperature=0.0,
-        )
+        response = None
+        # give it multiple chances to get a response
+        # if it fails, we'll try to condense memory
+        attempt = 0
+        while not response and attempt < self.attempts_to_condense:
+            # prepare what we want to send to the LLM
+            messages: list[dict[str, str]] = self._get_messages(state)
+            print('No of tokens, ' + str(self.llm.get_token_count(messages)) + '\n')
+            try:
+                if self.llm.is_over_token_limit(messages):
+                    raise TokenLimitExceededError()
+                response = self.llm.completion(
+                    messages=messages,
+                    stop=[
+                        '</execute_ipython>',
+                        '</execute_bash>',
+                        '</execute_browse>',
+                    ],
+                    temperature=0.0,
+                )
+            except (ContextWindowExceededError, TokenLimitExceededError):
+                # Handle the specific exception
+                print('An error occurred: ')
+                attempt += 1
+                # If we got a context alert, try trimming the messages length, then try again
+                if self.llm.is_over_token_limit(messages):
+                    # A separate call to run a summarizer
+                    self.condense(state=state)
+                    # Try step again
+                else:
+                    print('step() failed with an unrecognized exception:')
+                    raise ContextWindowLimitExceededError()
+
+            # TODO: Manage the response for exception.
         return self.action_parser.parse(response)
 
+    def condense(
+        self,
+        state: State,
```

Comment:
If you move this `.condense` function to the `LLM` class, you can make `messages` as the input argument. We might need to make each `Message` a Pydantic class with an additional attribute (e.g., `condensable`). This may also benefit the ongoing effort that tries to add vision capability in https://github.com/OpenDevin/OpenDevin/pull/2848#discussion_r1682039875  (cc @Kaushikdkrikhanu)

Reply:
This can be done. But on moving `.condense` to `LLM` class, it won't be able to add summary to `state.history` directly. I might also need to pass `state` as arg to `llm.completion` along with messages. 

If we are trying to make it cleaner (https://github.com/OpenDevin/OpenDevin/pull/2937#discussion_r1682877539), I instead was thinking of moving `.condense` to `MemoryCondenser` class, after making each `Message` a pydantic class. That's where I think `.condenser` should belong 😅.
WDYT? Which approach seems better?

Reply:
huh good point... maybe caller to `LLM` should make sure the summary got added to the `State` - like what we are doing now?

Reply:
On another thought: maybe `State` is not the place to add these new summarizing actions - `EventStream` should be: https://github.com/OpenDevin/OpenDevin/blob/c555fb68400bb117011cda3e5d3d95beb5169000/opendevin/controller/agent_controller.py#L395

Maybe you could add an optional argument for `LLM` class, like `condense_callback_fn`, which will be called with the condensed `Summarized` action whenever a condensation is happening. And that callback will add this `Summarize` action to the event stream?

Reply:
Can you elaborate on this? How does adding summarizing actions to EventStream help? 
MemGPT stores only the summary of events from the start to some `event.id`. We only need to store one summary in the history. A summary event remains useful until we create a new summary, after which it becomes useless.

Btw, I've moved the `.condenser` functions to `LLM` class. `MemoryCondenser` class used methods of `LLM` class, so I completely removed it and moved one of its method to `LLM` class. This also achieves the target (https://github.com/OpenDevin/OpenDevin/pull/2937#discussion_r1682877539)

Reply:
Basically, going forward, we hope to make `EventStream` a single source of truth for all our past message / history (i.e., you get all the history messages from the event stream), so ideally all your condensation actions also should goes into event stream.

But base on my understanding, your current implementation already emit "AgentSummarizeAction"? I think we just need to make sure that action got stored inside eventstream, we should be good to go on this bullet.

Reply:
@xingyaoww Oh interesting, I played with summaries in the event stream at some point. A couple of thoughts:
- are they 'real events'? That is, if emitted with the same mechanism as others (which they currently were not), would they be propagated downstream to components listening
- or, since they replace during runtime the real events they summarize, are they more like 'shadow events'? In which case they're not the truth, but a fleeting version of it, just necessary due to some runtime constraints; the source of truth holds the real events and is able to retrieve them when needed.

Ref: https://opendevin.slack.com/archives/C06QKSD9UBA/p1716737811950479?thread_ts=1716668475.888269&channel=C06QKSD9UBA&message_ts=1716737811.950479

- they need to be saved I assume, even if it's one at a time, to retrieve at a disconnect or restart/restore etc. Do we save them along with other events or we save them in State? If we save them in the eventstream, then we may want to consider moving some filtering logic too.

On a side note, making it only one in a run simplifies the logic either way.